tasks/README.md

```diff
@@ -59,9 +59,33 @@ The preferred style of the keys should be `snake_case`.
 
 **parameters**:  The parameters or input the task accepts listed with a [Puppet data type](../language/types_values_variables.md) string and optional description. Top level params names have the same restrictions as Puppet class param names and must match `\A[a-z][a-z0-9_]*\z`. Parameters may be `sensitive` in which case the task runner should hide their values in UI where possible.
 
-**implementations**: A list of implementation files in preference order, along with an optional list of required features for that implementation to be suitable on a target. The available features are defined by the task runner, but task runners should define at least the `shell`, `powershell` and `puppet-agent` features.
+**files**: Array of file names specified as strings that include the module containing the file. Files must be saved in module directories that Puppet makes available via mount points: files, lib, tasks. File specifications ending with `/` will require the entire directory. Files can also be provided as `implementations` objects. 
 
-#### Example
+**implementations**: A list of implementation objects. An implementation object describes resources and feature requirements that must be available on a target for specified resources to be utilized. The available features are defined by the task runner,  task runners should define at least the `shell`, `powershell` and `puppet-agent` features.
+
+When an executable is defined in a valid tasks directory it can be referenced by “name” in a hash with the following key value pairs.  
+**Key**: "name"  
+**Value**: executable name  
+**Type**: string  
+**Required**: True  
+
+**Key**: "requirements"  
+**Value**: required feature  
+**Type**: string  
+**Required**: False  
+
+`files` can be specified in `implementations` using the `files` key. When `files` are described using both the `files` and `implementations` metadata options the values are joined.  
```

Comment:
These sentences are right, but the specification below is wrong. There are 3 properties of an implementations object: `name` (required), `requirements` (optional), `files` (optional).

Also, this format of defining them is very verbose, I'd like something more compact.

Reply:
Supporting why I prefer a single object type: implementations are meant to be selected based on their requirements, the first object where the target matches all its requirements is selected. If we expect to match multiple objects with the same set of requirements, then walking this list gets a lot harder, and the use cases for what that's helpful are unclear to me. Generally if you have extra library files, they go with specific tasks, so it makes sense to group them together.